metadata_mapper/lambda_function.py

```diff
@@ -56,40 +75,37 @@ def map_page(payload, context):
     source_vernacular = vernacular_reader(payload)
     api_resp = source_vernacular.get_api_response()
     source_metadata_records = source_vernacular.parse(api_resp)
-    collection = payload.get('collection', {})
 
-    for enrichment_url in collection.get('rikolti__pre_mapping'):
-        enrichment_func, kwargs = parse_enrichment_url(enrichment_url)
-        if not enrichment_func and settings.SKIP_UNDEFINED_ENRICHMENTS:
-            continue
-        logging.debug(
-            f"[{collection['id']}]: running enrichment: {enrichment_func} "
-            f"for page {payload['page_filename']} with kwargs: {kwargs}")
-        source_metadata_records = [
-            record.enrich(enrichment_func, **kwargs)
-            for record in source_metadata_records
-        ]
+    source_metadata_records = run_enrichments(
+        source_metadata_records, payload, 'rikolti__pre_mapping')
 
     mapped_records = [record.to_UCLDC() for record in source_metadata_records]
     writer = UCLDCWriter(payload)
     if settings.DATA_DEST == 'local':
         writer.write_local_mapped_metadata(
             [record.to_dict() for record in mapped_records])
 
-    for enrichment_url in collection.get('rikolti__enrichments'):
-        enrichment_func, kwargs = parse_enrichment_url(enrichment_url)
-        if not enrichment_func and settings.SKIP_UNDEFINED_ENRICHMENTS:
-            continue
-        if enrichment_func in ['required_values_from_collection_registry',
-                               'set_ucldc_dataprovider']:
-            kwargs.update({'collection': collection})
-        logging.debug(
-            f"[{collection['id']}]: running enrichment: {enrichment_func} "
-            f"for page {payload['page_filename']} with kwargs: {kwargs}")
-        mapped_records = [
-            record.enrich(enrichment_func, **kwargs)
-            for record in mapped_records
-        ]
+    mapped_records = run_enrichments(
+        mapped_records, payload, 'rikolti__enrichments')
+    exceptions = {
+        rec.legacy_couch_db_id: rec.enrichment_report
+        for rec in mapped_records if rec.enrichment_report
+    }
+    if exceptions:
+        group_by_report = {}
+        for couch_id, reports in exceptions.items():
+            report = " | ".join(reports)
+            if report in group_by_report:
+                group_by_report[report].append(couch_id)
+            else:
+                group_by_report[report] = [couch_id]
+        count_by_report = {
+            report: f"{len(couch_ids)} of {len(mapped_records)}"
+            for report, couch_ids in group_by_report.items()
+        }
+        for report, count in count_by_report.items():
+            print(f"{count} records report enrichments errors: {report}")
```

Comment:
I've removed print statements that reported errors in each enrichment function and instead appended those errors to an array `enrichment_report` that exists on each record. This bit of logic here groups like lists, so if, for example, every single item on the page we're currently mapping has the exact same enrichment error, it's a little clearer to see. 

Rather than printing, this could get sent to the return value of map_page. We should maybe try to rationalize where to use printing vs. log messages vs. return values. I was trying to limit printing to command line utilities such as `map_registry_collections.py` and cases when execution should halt, rather than reporting something out and continuing to process. I was trying to only use log messages or return values in the internals such as this. I'm punting a bit on doing this rationalization until we have a clearer sense of our workflow management tooling.

Reply:
Sounds good. I have wondered about when to use print statements vs logging vs some other reporting method also. Figuring out a clear rationalization at some point sounds great.